crates/iota-transaction-builder/src/builder.rs

```diff
@@ -0,0 +1,1070 @@
+        self.command(Command::Upgrade(Upgrade {
+            modules: module.modules,
+            dependencies: module.dependencies,
+            package: package_id,
+            ticket: ticket.into_iter().next().unwrap(),
+        }));
+        name.push_named_commands(self);
+        self
+    }
+
+    /// Make a move vector from a list of elements.
+    pub fn make_move_vec<U: PTBArguments + MoveType>(
+        &mut self,
+        elements: impl IntoIterator<Item = U>,
+        name: impl NamedCommand,
+    ) -> &mut Self {
+        let mut args = Vec::new();
+        for e in elements {
+            args.extend(e.args(self));
+        }
+        let cmd = Command::MakeMoveVector(MakeMoveVector {
+            type_: Some(U::type_tag()),
+            elements: args,
+        });
+        self.commands.push(cmd);
+        name.push_named_commands(self);
+        self
+    }
+
+    /// Manually set a command with an optional name
+    pub fn named_command(&mut self, cmd: Command, name: impl NamedCommands) -> &mut Self {
+        self.commands.push(cmd);
+        name.push_named_commands(self);
+        self
+    }
+
+    async fn resolve_inputs(&mut self) -> Result<Vec<Input>, Error> {
+        let mut inputs = Vec::new();
+        for input in self.inputs.drain(..) {
+            let (object_id, mutable) = match input {
+                UnresolvedInput::ImmutableOrOwned(object_id)
+                | UnresolvedInput::Receiving(object_id) => (object_id, false),
+                UnresolvedInput::Shared { object_id, mutable } => (object_id, mutable),
+                UnresolvedInput::Resolved(input) => {
+                    inputs.push(input);
+                    continue;
+                }
+            };
+            let obj = self
+                .client
+                .object(object_id, None)
+                .await
+                .map_err(Error::Client)?
+                .ok_or_else(|| Error::Input(format!("missing object {object_id}")))?;
+
+            inputs.push(match obj.owner() {
+                Owner::Address(_) | Owner::Object(_) | Owner::Immutable => Input::ImmutableOrOwned(
+                    ObjectReference::new(object_id, obj.version(), obj.digest()),
+                ),
+                Owner::Shared(version) => Input::Shared {
+                    object_id,
+                    initial_shared_version: *version,
+                    mutable,
+                },
+            });
+        }
+        Ok(inputs)
+    }
+
+    async fn resolve_gas(&mut self) -> Result<(), Error> {
+        for gas_to_resolve in self.gas_to_resolve.drain(..) {
+            let obj = self
+                .client
+                .object(gas_to_resolve, None)
+                .await
+                .map_err(Error::Client)?
+                .ok_or_else(|| Error::Input(format!("missing object {gas_to_resolve}")))?;
+
+            let obj_ref = match obj.owner() {
+                Owner::Address(_) | Owner::Object(_) | Owner::Immutable => {
```

Comment:
Immutable also can't be used as gas object
Also it would probably be good to make sure that the gas objects match the gas owner address that's set `owner: self.sponsor.unwrap_or(self.sender),`

Reply:
Is it even possible to use an object owner for gas?